### ISSUE TITLE
ICONV issues for ne

### ISSUE THREAD

**Issue by mhosken** [hunspell/hunspell#366]
I have been asked to address the issue of spell checking Nepali in hunspell. To do this I need iconv rules. FYI the particular rules I am using are:
````
ICONV 5
ICONV <200C>_ <200C>_
ICONV <0930><094D><200C><092F> <0930><094D><200C><092F>
ICONV <0930><094D><200C><0935> <0930><094D><200C><0935>
ICONV <200C> <FFF0>
ICONV <200D>_ _
````

You will see that I am running quite the gamut of iconv behaviour, particularly needing _ support, which I read to mean string end constraint and the ability to delete. The rules as stated delete in two ways, the ICONV <200C> <FFF0> does a generic delete by mapping to a character that is then IGNOREd. The ICONV <200D>_ _ does a constrained mapping. As I start thinking about how to implement these, I am thinking that the _ has no place in the right hand side of a rule and that the rule set should become:

````
ICONV 5
ICONV <200C>_ <200C>
ICONV <0930><094D><200C><092F> <0930><094D><200C><092F>
ICONV <0930><094D><200C><0935> <0930><094D><200C><0935>
ICONV <200C> <FFF0>
ICONV <200D>_ <FFF0>
````

As to implementation. Even without considering string boundary constraints, there is a bug in RepList::near() where strcmp() should be replaced by strncmp based on strlen(dat[m]->pattern).

Another problem is that RepList::conv() iterates a UTF-8 string by byte rather than character. Not sure what to do about that one and it's not a huge disaster since each match string will always start with a UTF-8 start byte, so iterating by bytes merely wastes a bunch of time.

I am thinking of diving in and addressing these and the string boundary constraints issue (via multiple output patterns for each of the 4 possible boundary constraint conditions: none, left, right, both). I would probably also merge RepList::near() and RepList::match().

Am I on the right track? Would you be interested in such a PR?

**Comment by mhosken** [comment]
There is a further issue. As it stands, by sorting the input pattern strings and doing a binary search to find the right one, it is impossible to address the situation where one pattern is a prefix to another pattern. For example, consider the 3 input patterns: `abc`, `abd`, `ab`. These would be stored as `ab`, `abc`, `abd`. For an input string of `abc`, it would not be possible to ensure that the binary search would match the `abc` pattern over the `ab` pattern.

The alternatives for a solution to this are:

* Change the storage order of the strings so that short prefixes are stored after the longer strings and then linearly search.
* Have two replacement objects. One is a simple object with an input pattern and 4 output patterns (for the 4 string contexts). The other is a complex object containing a list of simple replacements that are linearly searched, stored in inverse length order.

A further alternative is just to document the problem and hope nobody has a real use case that requires it. At least we know what the problem is and can indicate it and wait for a feature request once such a real use case emerges.

**Comment by dimztimz** [comment]
The underscore thingy should be part of suffix stripping, not of ICONV.

**Comment by dimztimz** [comment]
ICONV OCONV should behave like the second variant of [strtr in PHP](http://php.net/manual/en/function.strtr.php).

Related to #480

**Comment by changwoo** [comment]
How about just changing the longer-pattern searching code to scan forward, rather than backward?

```diff
diff --git a/src/hunspell/replist.cxx b/src/hunspell/replist.cxx
index 89d4caa..74addb7 100644
--- a/src/hunspell/replist.cxx
+++ b/src/hunspell/replist.cxx
@@ -109,8 +109,8 @@ int RepList::find(const char* word) {
       p2 = m - 1;
     else if (c > 0)
       p1 = m + 1;
-    else {      // scan back for a longer match
-      for (p1 = m - 1; p1 >= 0; --p1)
+    else {      // scan forward for a longer match
+      for (p1 = m + 1; p1 <= p2; ++p1)
         if (!strncmp(word, dat[p1]->pattern.c_str(), dat[p1]->pattern.size()))
           m = p1;
         else if (dat[p1]->pattern.size() < dat[m]->pattern.size())
```

**Comment by changwoo** [comment]
I also think such Nepali feature can be implemented using suffix rules.

If this understore feature is to be kept, at least it should be documented.

**Comment by mhosken** [comment]
Does your patch work? By my understanding, it shouldn't since I go to some work to try to order entries by length first so that one would get the ordering: abd abc ab acd ac a. But I might have got that wrong.

As to using suffix rules to support Nepali, could you explain that further? From my understanding, it's going to be ugly at the very least. 200C is ignored except at the ends of words and the 2 special cases. Whereas 200D is only ignored word finally. It's marginal whether the 2 special cases can be handled by suffixation, but the global ignoring doesn't seem to fit well in my understanding, with suffixation.

**Comment by changwoo** [comment]
Yes, the patch works. The order of the entries are not in that longest-first way. Anyway, why are these entries sorted in that way if they are not intended to be linear-searched?

**Comment by changwoo** [comment]
I stand corrected about the suffix rules. I know what U+200D (ZWJ) and U+200C (ZWNJ) mean. It is not possible to implement ZWJ/ZWNJ with SFX rules.

**Comment by dimztimz** [comment]
The `ICONV <200C>_ <200C>`

This rule says that if we have a string that ends with <200C> as the before last char, and then with arbitrary character, strip the very last character. This is indeed suffix stripping, but the current suffix stripping requires the exact suffix to be known.

The current rules are SFX FLAG BASE_WORD_STRIP DERIVED_WORD_ADD

Both BASE_WORD_STRIP and DERIVED_WORD_ADD have to be fully defined, no arbitrary characters there.

Internally it is done the other way, we strip the suffix from the derived input word, we add the base word suffix, practically stemming it, and then we check if the stemmed word is inside the dictionary. But that is not important now.

For V2 we can add a new system separate from ICONV and suffix stripping, it will be regex substitutions which should serve the Nepali needs.

**Comment by mhosken** [comment]
No this rule says: if 200C occurs at the end of a word, keep it. But if it occurs anywhere inside the word, convert it to the ignored FFF0 (done by another ICONV rule), in effect deleting it. This isn't something a suffix can do.

**Comment by changwoo** [comment]
Please document this feature in hunspell(5).

**Comment by shantanuo** [comment]
Has this feature properly documented?
It is important not just for Nepali Language but all Indian languages.

**Comment by ssvb** [comment]
This undocumented underscores feature in the ICONV directives sort of gets in the way of https://github.com/hunspell/hunspell/pull/1118
Can somebody step up and finally document it? What exactly is needed there for Nepali Language and Indian languages?

**Comment by shantanuo** [comment]
The Marathi affix file currently contains the following rule:

`ICONV ं_ े`

This rule is intended to ensure that if an Anusvara (Unicode character U+0902) appears at the end of a word, it is replaced with े (Unicode character U+0947). If the Anusvara appears elsewhere, it should remain unchanged and be ignored.
This transformation is particularly important for Marathi because many entries in the .dic file end with े But some users type the same words using a final Anusvara instead. Without support for the underscore-based `ICONV` rule, it would be necessary to duplicate all such dictionary entries by creating additional variants ending with Anusvara. This would considerably increase the size of the `.dic` file.
Therefore, support for this feature is important for Marathi language processing and dictionary optimization.

**Comment by ssvb** [comment]
Thanks for a very quick response.

> The Marathi affix file 

Could you additionally help me to find a Marathi dictionary for Hunspell with such affix file for testing? That would be very much appreciated.

> currently contains the following rule:
> 
> `ICONV ं_ े`
> 
> This rule is intended to ensure that if an Anusvara (Unicode character U+0902) appears at the end of a word, it is replaced with े (Unicode character U+0947). If the Anusvara appears elsewhere, it should remain unchanged and be ignored.

Do you happen to know why `_` was picked for this syntax? I mean, instead of `^` and `$`, that would be consistent with the syntax of REP entries.

BTW, is there any way to escape the '_' character itself? If not, then why not?

> Therefore, support for this feature is important for Marathi language processing and dictionary optimization.

Would you volunteer to create a pull request with the documentation update? Or at least write a block of text exactly in the way how it should to be added to https://github.com/hunspell/hunspell/blob/master/man/hunspell.5

Also the status of this issue is still "open". Does this mean that something is still not working correctly for Nepali?

**Comment by shantanuo** [comment]
>> Could you additionally help me to find a Marathi dictionary for Hunspell with such affix file for testing? 

https://extensions.libreoffice.org/en/extensions/show/marathi-spellchecker

>> Do you happen to know why _ was picked for this syntax? I mean, instead of ^ and $, that would be consistent with the syntax of REP entries.

I agree that using $ would appear to be a more logical and regex-consistent choice. However, I do not know the reason why _ was originally selected for this syntax.

>> BTW, is there any way to escape the '_' character itself? If not, then why not?

I have no idea about that.

>> Would you volunteer to create a pull request with the documentation update? 

I have already created the pull request for the documentation update: 
https://github.com/hunspell/hunspell/pull/964

>> Also the status of this issue is still "open". 

In my opinion, the status of this issue should now be changed to "closed".
    
>> Does this mean that something is still not working correctly for Nepali?

Everything is currently working as expected for both Nepali and Marathi.

**Comment by ssvb** [comment]
> https://extensions.libreoffice.org/en/extensions/show/marathi-spellchecker

Thanks! What's the difference between `mr_IN.aff` / `mr_IN.dic` and `mr_IN_eng.aff` / `mr_IN_eng.dic` inside of this package after it is unzipped?

> I have already created the pull request for the documentation update: [#964](https://github.com/hunspell/hunspell/pull/964)

Thank you. That's a good start. However the documentation is still:
1. incomplete, as it describes only matching the end of the word, but the Hunspell code seems to also have the matching logic for the beginning of the word too.
2. might be somewhat misleading, as it shows hexadecimal numbers in angled braces `"ICONV <200D>_ <FFF0>"`, which could be possibly mistakenly confused with Unicode codepoints.
 
But in reality, using

test.aff:
```
SET UTF-8
WORDCHARS <0123456789ABCDEF>
ICONV 1
ICONV <200D> <FFF0>
```

test.dic:
```
1
<FFF0>
```

Results in:
```
$ hunspell -d test
Hunspell 1.7.2
<FFF0>
*

<200D>
*
```
Basically, both `"<FFF0>"` and `"<200D>"` are just 6-character-long words, assuming that the tokenizer does not break them (the WORDCHARS directive prevents this).

**Comment by ssvb** [comment]
> Everything is currently working as expected for both Nepali and Marathi.

By the way, I'm developing https://github.com/ssvb/hunaftool, which is basically a better replacement for the `munch` and `unmunch` tools from Hunspell. But right now using `hunaftool` to unpack `mr_IN.aff + mr_IN.dic` to a word list results in these unpacked words not getting recognized by Hunspell. My preliminary analysis hints that the tokenizer  likely breaks these unpacked words into parts and these chopped parts are rejected. The affix file contains some complicated ICONV and BREAK directives, which I'm unable to understand as I don't speak Marathi and don't know anything about it. Ideally, I would want my tool to be able to correctly decode/encode any Marathi dictionary.

Here's a reduced `.dic` file:
```
1
अंकलिपी/a
```

It gets decoded into the following 3 words:
```
अंकलिपी
अंकलिपीनं
अंकलिपीने
```

Do these three decoded Marathi words look correct to you? Hunspell rejects them (after apparently splitting each of them into two pieces):
```
$ hunspell -d mr_IN -l reduced-wordlist.txt 
अंकल
प
अंकल
प
अंकल
प
```

Is the dictionary probably missing a correct WORDCHARS directive? How do you see that it works as expected? Are you testing the dictionary with LibreOffice only, disregarding the Hunspell's command line tool?

**Comment by shantanuo** [comment]
>> Do these three decoded Marathi words look correct to you? 

Yes

>> Are you testing the dictionary with LibreOffice only, disregarding the Hunspell's command line tool?

Yes

>> it describes only matching the end of the word, but the Hunspell code seems to also have the matching logic for the beginning of the word too.

Good point. However, I was looking at it primarily from my use-case perspective. Please feel free to improve the documentation accordingly.

As per my understanding, `ICONV` processing happens before the `BREAK` rules are applied.

I think this discussion is now going off-topic. If you have any further questions, please feel free to contact me at shantanu dot oak at gmail dot com

**Comment by ssvb** [comment]
> Good point. However, I was looking at it primarily from my use-case perspective. Please feel free to improve the documentation accordingly.

Yes, I understand that. I'm going to add more testcases for all edge cases of the current ICONV implementation and more detailed documentation for it in a pull request.

My only concern is whether all of this (partially undocumented) functionality is really desired in Hunspell. Supporting something that is not used by the real dictionaries is just a liability in my opinion. That's why I'm asking questions.

> As per my understanding, `ICONV` processing happens before the `BREAK` rules are applied.
> I think this discussion is now going off-topic. If you have any further questions, please feel free to contact me at shantanu dot oak at gmail dot com

Do you mean that the suspected tokenizer differences between LibreOffice and the Hunspell's command line tool are offtopic here? Rather than communicating privately, I would prefer to open a public bugreport. But I don't know whether the Marathi dictionary has a bugtracker of its own.

**Comment by ssvb** [comment]
> Here's a reduced `.dic` file:
> 
> ```
> 1
> अंकलिपी/a
> ```
> 
> It gets decoded into the following 3 words:
> 
> ```
> अंकलिपी
> अंकलिपीनं
> अंकलिपीने
> ```
> 
> Do these three decoded Marathi words look correct to you? Hunspell rejects them (after apparently splitting each of them into two pieces):
> 
> ```
> $ hunspell -d mr_IN -l reduced-wordlist.txt 
> अंकल
> प
> अंकल
> प
> अंकल
> प
> ```
> 
> Is the dictionary probably missing a correct WORDCHARS directive?

BTW, this particular problem in a reduced testcase has been just resolved by @caolanm via https://github.com/hunspell/hunspell/commit/5aef30828075aa5376d54f3267ac343c703fc2dd.

There seem to be still some other decoded Marathi words that are not recognized by the Hunspell's command line tool. However, since ["Everything is currently working as expected for both Nepali and Marathi"](https://github.com/hunspell/hunspell/issues/366#issuecomment-4404162917) (as far as the ICONV functionality is concerned), investigating this further does not belong here.

**Comment by caolanm** [comment]
Your own 

commit 73b1cad1af7ab94252f75784fa6724cf062a6966
Date:   Mon Apr 18 16:28:26 2016 +0700

    Add support for bounded conversion

fixed the substantive issue. There were other related changes of 

commit d7045b6ed7466b375b59a877d2129d2558beacc5
Date:   Wed Mar 22 15:16:15 2017 +0100

    Fix RepList::find so it matches the longest, issue #480

commit 25508749e99c34d047dbc204ff3f48f52d3f52b1
Date:   Wed Mar 15 15:26:09 2017 +0900

    Correct and simplify Replist::add and Replist::find

commit e5ce1835a012bca70a214a06c77df1f440166570
Date:   Sat May 30 13:44:32 2026 +0000

    Resolves #1126 retry shorter ICONV/REP keys when the longest is anchored

So I think we can close this as effectively done
